DatasourceEditor: Add UI to edit custom HTTP headers

adrien-f opened the pull request:
Support for custom headers was added in #11643 but was missing in the UI.

Fixes #12779

---

This is a WIP to gather implementation and design feedback from the team. Here's how it looks like for now:

![image](https://user-images.githubusercontent.com/781688/60434607-cd06f480-9c07-11e9-88dc-de6b1a09f1cd.png)

Right now I'm trying to brainstorm the best way to work with the format of the custom headers with `httpHeaderName$INDEX$` and `httpHeaderValue$INDEX$` which doesn't make it easy. Suggestions welcomed !

adrien-f committed:
DatasourceEditor: Add UI to edit custom HTTP headers

Support for custom headers was added in #11643 but was missing in the UI.

Fixes #12779

adrien-f commented:
Also trying to sign the CLA but the assistant doesn't work well right now.

marefr commented:
@adrien-f this looks like a great start and UI wise looks smooth. Not sure if you have any react experience as @torkelo requested you to rewrite in react? However, I'll let someone add some pointers here regarding how to convert your CustomerHeaderSettingsCtrl to React.

adrien-f commented:
👋 Thanks for the feedback you two ! I'm not sure I can only convert the CustomHeaderSettingsCtrl to React as the page around it is Angular ? 

I would be okay helping you migrate the whole datasource editing page to React but that feels like a big undertaking. Do you already have some WIP on this ? If not, I can start iterating on this.

Cheers

marefr commented:
> I'm not sure I can only convert the CustomHeaderSettingsCtrl to React as the page around it is Angular

You can. @dprokop will provide more information about how

torkelo commented:
Almost everything on the data source settings page is already in React it's just the http settings that are still in angular. 

https://github.com/grafana/grafana/blob/master/public/app/features/datasources/settings/PluginSettings.tsx

dprokop commented:
@adrien-f thanks a lot for your contribution!

We provide a way to use React components from within angular templates. A good starting point for you would be taking a look at https://github.com/grafana/grafana/blob/master/public/app/core/angular_wrappers.ts which a file used for defining the bridges between Angular and React. 

the `react2AngularDirective` takes 3 arguments:
1. Name of the directive. This is specified in camelCase. Later on it's used in angular templates with kebab-case. In your case its gonna be `datasourceCustomHeadersSettings`
2. React component to render. This is is the "magic" piece. You basically specify your react component that's gonna be rendered when the directive is used.
3. Array of attributes that you will pass from Angular to to React. Changes in those atrributes will also make the React component to re-render if necessary. 

For reference usage you can take a look how we implemented `secret-form-field` React component:)

As a starting point I suggest moving the CustomHeadersSettingsCtrl logic into HttpSettingsCtrl and implementing the custom_headers_settings form itself in React.

Let me know if anything is not clear!

adrien-f commented:
Thanks @dprokop for the tips ! I've managed to make the bridge work 👍 

I'll keep working on this this weekend and hopefuly some time next week as well.

adrien-f commented:
It's me again 😄 !

Basic logic is here and working, you can add, reset and remove headers. There's a few catches with the way secureJsonData is handled so I did not manage to delete a header value once set ?

Also the fact that the index in `useCustomHeaders` starts at `1` adds a bit of index offsets to deal with.

A design issue that still needs to be solved, help would be appreciated:

![image](https://user-images.githubusercontent.com/781688/61632720-d545d300-ac8d-11e9-9b75-4ab0d3fc97d1.png)

adrien-f commented:
👋 Tests have been done, I'll work on getting the CI green and I'd love a first round of reviews/feedback.

Thanks again !

adrien-f commented:
Hiyo ! Back from my holidays, let me know when you have some time for review 😃

dprokop commented on public/app/features/datasources/settings/CustomHeadersSettings.tsx:
```diff
@@ -0,0 +1,174 @@
+
+export interface Props {
+  datasource: {
+    jsonData: {
```
If you merge latest master you can use `KeyValue<T>` from @grafana/data package to type those props.

import { KeyValue } from '@grafana/data'

dprokop commented on public/app/features/datasources/settings/CustomHeadersSettings.tsx:
```diff
@@ -0,0 +1,174 @@
+    this.updateHeaders();
+  }
+
+  onHeaderChange(
```
I suggest refactoring this method into 4 separate ones:

- `onHeaderAdd`
- `onHeaderChange`
- `onHeaderReset`
- `onHeaderRemove`

It will improve the readability of this logic. You will be able to remove the `resetHeader` and `updateHeader` functions and add those implementations respectively in `onHeaderReset`, `onHeaderChange`

dprokop commented on public/app/features/datasources/settings/CustomHeadersSettings.tsx:
```diff
@@ -0,0 +1,174 @@
+            return { headers: [...prevState.headers, { name: 'X-Custom-Header', value: '', configured: false }] };
+          case HeaderChangeEvent.update:
+            const { name, value } = eventTarget as EventTarget & HTMLInputElement;
+            return { headers: prevState.headers.map(this.updateHeader(headerIndex, name, value)) };
```
The `name` param in `updateHeader` does not make much sense as it's always 'value'. It should be enough to just pass index and the value.

Also, you should not rely on eventTarge.name, as this might be a subject of change, that won't be picked up by Typescript, and may lead to bugs.

dprokop commented on public/app/features/datasources/settings/CustomHeadersSettings.tsx:
```diff
@@ -0,0 +1,174 @@
+            };
+        }
+      },
+      () => this.props.onChange(this.state.headers)
```
I think you should update the controller on input blur event instead of every change as until then you can think about the value as a transient state

dprokop commented on public/app/features/datasources/settings/HttpSettingsCtrl.ts:
```diff
@@ -24,6 +24,27 @@ coreModule.directive('datasourceHttpSettings', () => {
+          }
+
+          Object.keys($scope.current.jsonData)
+            .sort()
```
Any particular reason for sorting here? Header names are added with incremental index, so when retrieving the keys from `jsonData`, the ones that you are interested in should be sorted already.

dprokop commented:
@adrien-f left some review comments for you. Also, I have noticed, that the remove button has some problems with alignment:

![image](https://user-images.githubusercontent.com/2376619/63919642-723f1d00-ca3f-11e9-951e-99a4758e11a7.png)


Apart from that, it looks great!

adrien-f commented:
Thanks you much for the review ! Will get it on this weekend/early-week 🚀

marefr commented:
@adrien-f is this ready for review again?

adrien-f commented:
I just need to finish tweaking the CSS alignment for the X button, not my best skill 😅

zeeraw commented:
Haha, I'm so excited about this 😅

dprokop commented:
@adrien-f is this ready for another review?

JarleB commented:
+1 for this. Any news ?

adrien-f commented:
I'm currently refactoring that MR to adapt to the new React page it's in. This should simplify things a lot.

adrien-f commented:
So I've pushed the latest changes:

* The delete button is now properly aligned
* But the reset button not anymore since I rebased, is this on your end ? I'll try to fix it
* Moved the React components to `packages/grafana-ui`

There appears to be no `Jest` support in this new part yet? How would you recommend I go from there?

Thanks again for the review, we're nearly there !

adrien-f commented:
Tests are green now :)

dprokop commented on packages/grafana-ui/src/components/SecretFormFied/SecretFormField.tsx:
```diff
@@ -5,7 +5,7 @@ import { FormField } from '../FormField/FormField';
   // Function to use when reset is clicked. Means you have to reset the input value yourself as this is  uncontrolled
   // component (or do something else if required).
-  onReset: () => void;
+  onReset: (e: any) => void;
```
I don't think you need this

dprokop commented on packages/grafana-ui/src/components/DataSourceSettings/CustomHeadersSettings.tsx:
```diff
@@ -0,0 +1,198 @@
+                  labelWidth={5}
+                  inputWidth={header.configured ? 11 : 12}
+                  placeholder={'Header Value'}
+                  onReset={_ => this.onHeaderReset(index)}
```
Instead of using `_` I suggest doing `() => ...`. This will remove the necessity for `onReset` type signature change in `SecretFormField`

dprokop commented on packages/grafana-ui/src/components/DataSourceSettings/CustomHeadersSettings.tsx:
```diff
@@ -0,0 +1,198 @@
+  }
+
+  updateSettings(headers: CustomHeaders) {
+    const { jsonData, secureJsonData } = this.props.dataSourceConfig;
```
`secureJsonData` can be undefined, hence I am getting this error when adding first header:

![image](https://user-images.githubusercontent.com/2376619/68937067-0ba6e100-079c-11ea-8b24-d6f7715b3a48.png)

dprokop commented on packages/grafana-ui/src/components/DataSourceSettings/CustomHeadersSettings.tsx:
```diff
@@ -0,0 +1,198 @@
+    headers: [],
+  };
+
+  constructor(props: Props) {
```
Unnecessary constructor

dprokop commented on packages/grafana-ui/src/components/DataSourceSettings/CustomHeadersSettings.tsx:
```diff
@@ -0,0 +1,198 @@
+    const { jsonData, secureJsonData } = this.props.dataSourceConfig;
+
+    for (const [index, header] of headers.entries()) {
+      jsonData[`httpHeaderName${index + 1}`] = header.name;
```
`httpHeaderName` -> can you pls extract it to a const?

dprokop reviewed (changes requested):
@adrien-f once again thanks for this contribution. I went through the code and left some minor comments. I have also found one bug(left a comment too).

This is looking very close to mergeable. Please apply the changes and we should be good to merge this:)

dprokop commented:
> There appears to be no Jest support in this new part yet? How would you recommend I go from there?

what do you mean by that @adrien-f ?

dprokop commented:
@adrien-f hey! Do you plan to work on this soon?

marefr commented:
@dprokop probably time for you to pick this up

dprokop committed:
Merge branch 'master' into feature/datasources-ui-headers

dprokop committed:
Review

dprokop committed:
Layout updates

dprokop commented:
@marefr @torkelo if you don't mind, please review this. I've picked this up and refactored/updated look a bit.

torkelo reviewed (approved):
Looks good! parsing out there values and updating the state feels a bit too complex but not sure exactly how to improve that :)

peterholmberg reviewed (approved):
lgtm

marefr commented:
@dprokop moved it to 6.7 milestone
